substrate/frame/referenda-tracks/src/lib.rs

```diff
@@ -29,99 +46,172 @@ mod tests;
 	pub enum Event<T: Config<I>, I: 'static = ()> {
-		// Foo(T::AccountId),
+		/// A new track has been inserted
+		Created { id: TrackIdOf<T, I> },
+		/// The information for a track has been updated
+		Updated { id: TrackIdOf<T, I>, info: TrackInfoOf<T, I> },
+		/// A track has been removed
+		Removed { id: TrackIdOf<T, I> },
 	}
 
 	#[pallet::error]
-	pub enum Error<T, I = ()> {}
+	pub enum Error<T, I = ()> {
+		/// The maxmimum amount of track IDs was exceeded
+		MaxTracksExceeded,
+		/// The specified ID for this track was not found
+		TrackIdNotFound,
+		/// The specified ID for this track was already existing
+		TrackIdAlreadyExisting,
+		/// The track cannot be removed
+		CannotRemove,
+	}
+
+	#[pallet::call(weight(<T as Config<I>>::WeightInfo))]
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// Insert a new referenda Track.
+		///
+		/// Parameters:
+		/// - `id`: The Id of the track to be inserted.
+		/// - `info`: The configuration of the track.
+		/// - `pallet_origin`: A generic origin that will be matched to the track.
+		///
+		/// Emits `Created` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[pallet::call_index(0)]
+		pub fn insert(
+			origin: OriginFor<T>,
+			id: TrackIdOf<T, I>,
+			info: TrackInfoOf<T, I>,
+			pallet_origin: PalletsOriginOf<T>,
+		) -> DispatchResultWithPostInfo {
+			T::UpdateOrigin::ensure_origin(origin)?;
+
+			ensure!(Tracks::<T, I>::get(id) == None, Error::<T, I>::TrackIdAlreadyExisting);
+
+			TracksIds::<T, I>::try_append(id).map_err(|_| Error::<T, I>::MaxTracksExceeded)?;
+			Tracks::<T, I>::set(id, Some(info));
+			OriginToTrackId::<T, I>::set(pallet_origin.clone(), Some(id));
+
+			Self::deposit_event(Event::Created { id });
+			Ok(().into())
+		}
+
+		/// Update the configuration of an existing referenda Track.
+		///
+		/// Parameters:
+		/// - `id`: The Id of the track to be updated.
+		/// - `info`: The new configuration of the track.
+		///
+		/// Emits `Updated` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[pallet::call_index(1)]
+		pub fn update(
+			origin: OriginFor<T>,
+			id: TrackIdOf<T, I>,
+			info: TrackInfoOf<T, I>,
+		) -> DispatchResultWithPostInfo {
+			T::UpdateOrigin::ensure_origin(origin)?;
+
+			Tracks::<T, I>::try_mutate(id, |track| {
+				if track.is_none() {
+					return Err(Error::<T, I>::TrackIdNotFound);
+				};
+
+				*track = Some(info.clone());
+
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::Updated { id, info });
+			Ok(().into())
+		}
+
+		/// Remove an existing track
+		///
+		/// Parameters:
+		/// - `id`: The Id of the track to be deleted.
+		///
+		/// Emits `Removed` event when successful.
+		///
+		/// Weight: `O(n)`
+		#[pallet::call_index(2)]
+		pub fn remove(origin: OriginFor<T>, id: TrackIdOf<T, I>) -> DispatchResultWithPostInfo {
+			T::UpdateOrigin::ensure_origin(origin)?;
+
+			ensure!(Tracks::<T, I>::contains_key(id), Error::<T, I>::TrackIdNotFound);
+
+			OriginToTrackId::<T, I>::iter_keys().for_each(|origin| {
```

Comment:
Iterating all the origins looks unnecessarily expensive, I think it's reasonable to expect the extrinsic to be called with the corresponding pallet_origin.

Reply:
I was thinking the same.

However, I noticed this structure allowed for a multiple origins to single track setup, like the one that exists on OpenGov currently.

My guess is that, unless we figure out how to make that setup work, this solution has a low chance of being accepted, as it would turn out to be a worse solution than what's already out there (?)

Maybe I'm wrong on the last part, but it's worth exploring those scenarios, and in that case, this expense is necessary.